### DISCUSSION TITLE
Undefined behavior for non-idempotent slice selectors?

### DISCUSSION THREAD

**Discussion by tiagobento** [pmndrs/zustand#2182] [Bug report] [answered]
Hi everyone! First of all, thanks a lot for this amazing project. I'm using Zustand with Immer and it's been a delight! 

I'm using Zustand on a fairly big component rewrite, and my browser tab was crashing when I changed something very specific, seemingly unrelated.

This is the commit that fixed it:

https://github.com/tiagobento/kie-tools/commit/c3c0389a95be00dc8cccf1fe218b03528c3b93fa

Apparently, the selector for `useStoreWithEqualityFn` (2nd parameter) can't be passed a function that doesn't behave "idempotently".

This is probably a very dumb thing that I did, but anyway, there's nothing stopping anyone from doing that and the consequences are very severe... Took me ~4h to figure it out. I took everything apart before I found the culprit.
 
 Have you seen this before? I couldn't find anything remotely related (maybe I'm not that experienced with Zustand yet to search accurately).

**Comment by dai-shi** [comment, marked as answer]
I wouldn't say it's an undefined behavior. It should be just extra re-renders. Not sure how it causes such a crash.

It's somewhat a well-known limitation with selectors and hooks in general, but it can be new to some users.

I would prefer to fix it like this:
```js
const EMPTY_ARRAY = [];
export function IncludedModels() {
  const dmnEditorStoreApi = useDmnEditorStoreApi();
  const thisDmnsImports = useDmnEditorStore((s) => s.dmn.model.definitions.import ?? EMPTY_ARRAY);
```

**Comment by tiagobento** [comment]
@dai-shi Hi! Thanks for the quick reply. I can confirm that your solution works as well. It makes total sense, and it looks like my error was caused by this extra re-renders unchaining a series of state updates that ultimately resulted in a crash of the browser tab. I've had many infinite loops in the past with useEffects and whatnot, but this was the first time my browser tab was completely irresponsive. 

Anyway, I consider this solved and am glad to hear that it is not a bug or a unknown undefined behavior in Zustand. Thx a lot!
